### PR TITLE
Renamed some stuff. Added server message. Closes tcp connection.

### PR DESCRIPTION
**Pull request by Ploug**
Added alert when server is down.

### CODE REVIEW

File: src/application/communication/Broker.java [hunk 1]
```diff
@@ -506,17 +527,8 @@ private void readDestroyedObject(ByteBuffer input)
...
-     * @param input
-     *            The ByteBuffer that handles reading of data send from the server.
-     */
-    private void readNewPlayer(ByteBuffer input) // Should probably tell
-                                                 // GameClient about the new
-                                                 // player instead
+    private String readString(ByteBuffer input)
     {
-        GameObjectDAO data = new GameObjectDAO();
         int nicknameLength = (int)input.get();
```
Comment (LuqJensen) [inline src/application/communication/Broker.java]: if this is used multiple places. the variable name is no longer legit

File: src/application/communication/GameClient.java [hunk 2]
```diff
@@ -163,7 +185,13 @@ private void joinGame(Pane gameBox, Specializations spec)
         System.out.println("trying to join game");
         try
         {
-            Broker broker = new Broker();
+            if(broker != null)
+            {
+                broker.unregisterAll(this);
```
Comment (LuqJensen) [inline src/application/communication/GameClient.java]: shouldnt this be done in deactivate?

File: src/application/communication/GameClient.java [hunk 3]
```diff
@@ -151,6 +158,21 @@ public void setSceneRelativeLocation(TranslatedPoint sceneRelativeLocation)
         }
 
     }
+    private void brokerException()
```
Comment (LuqJensen) [inline src/application/communication/GameClient.java]: never name a method Exception, that word is reserved for Exceptions.
name it brokerError or brokerFault or something else

**Comment by LuqJensen** [comment]
We should also have a finalize method on broker so that we are sure the connections will be cleaned up no matter what

File: src/application/communication/GameClient.java [hunk 4]
```diff
@@ -35,15 +37,17 @@
     private Account account;
     private CharacterController characterController;
     private TranslatedPoint sceneRelativeLocation;
+    private Broker broker;
```
Comment (LuqJensen) [inline src/application/communication/GameClient.java]: broker is part of the map, is this necessary?..

File: src/application/engine/game_object/View.java [hunk 5]
```diff
@@ -1,6 +1,7 @@
 package application.engine.game_object;
 
 import application.engine.entities.BoDCharacter;
+import application.engine.entities.Bullet;
```
Comment (LuqJensen) [inline src/application/engine/game_object/View.java]: ?

File: src/application/engine/rendering/ClientMap.java [hunk 6]
```diff
@@ -141,13 +141,11 @@ public void activate()
...
             leaderboard.refresh();
-            double total = 0;
-            for (double d : drawtimes)
-            {
-                total += d;
-            }
-            fps = (int)(1 / (total / drawtimes.size()));
-            drawtimes.clear();
+        });
+        LightEvent fpsEvent = new LightEvent(1000, () ->
```
Comment (LuqJensen) [inline src/application/engine/rendering/ClientMap.java:145]: nice (y)

File: src/application/engine/rendering/ClientMap.java [hunk 7]
```diff
@@ -224,13 +221,19 @@ public void handle(long currentNanoTime)
...
+                gc.setFill(Color.BLACK);
+                gc.fillText(serverMessage, canvas.getWidth()/2, canvas.getHeight()-200);
+                gc.setFont(Font.font("Verdana", 12));
+                ++frames;
                 canvas.requestFocus();
             }
         };
         animationTimer.start();
-
+//        for(String s : Font.getFontNames())
```
Comment (LuqJensen) [inline src/application/engine/rendering/ClientMap.java]: remove this

File: src/application/engine/rendering/ClientMap.java [hunk 8]
```diff
@@ -485,4 +489,23 @@ public void setScale(double xFactor, double yFactor)
         gc.setTransform(startingAffine);
         gc.scale(xFactor, yFactor);
     }
+
+    public void writeServerMessage(String readString)
+    {
+        new Thread(() ->
```
Comment (LuqJensen) [inline src/application/engine/rendering/ClientMap.java:498]: we should instead have a messagepump with lightevents. on each update we could then iterate its elements and they would remove themselves if they have timedout

Comment (LuqJensen) [inline src/application/engine/rendering/ClientMap.java:498]: same goes for kill notifications

File: src/application/communication/GameClient.java (hunk 2 above)
Comment (Ploug) [inline src/application/communication/GameClient.java]: unregisterAll, unregisters gameclient as a observer from all observations right? There is no deactivate in gameclient, so what deactivate are you talking about?

Comment (LuqJensen) [inline src/application/communication/GameClient.java]: i see your point. however i think gameclient should stop observing broker when it deactivates. so maybe that functionality must be called from broker

File: src/application/engine/rendering/ClientMap.java (hunk 8 above)
Comment (Ploug) [inline src/application/engine/rendering/ClientMap.java:498]: Not entirely sure what you want here.

File: src/application/communication/Broker.java [hunk 9]
```diff
@@ -277,7 +297,8 @@ private void receiveUdp()
                 }
                 catch (IOException e)
                 {
-                    e.printStackTrace();
+
+                    notifyObservers(Observation.SERVER_OFFLINE);
```
Comment (LuqJensen) [inline src/application/communication/Broker.java:301]: this isnt necessarily a server issue

File: src/application/communication/Broker.java [hunk 10]
```diff
@@ -417,6 +438,7 @@ private void startListeningTCP()
             }
             catch (IOException e)
             {
+                this.notifyObservers(Observation.SERVER_OFFLINE);
```
Comment (LuqJensen) [inline src/application/communication/Broker.java:441]: this isnt necessarily a server issue

**Pull request closed without merging**